### PR TITLE
improve server testing

### PR DESCRIPTION
**Pull request by eed3si9n**
previously I was using separate thread in a forked test to test the server, but that is not enough isolation to run multiple server tests.
This adds `RunFromSourceMain.fork(workingDirectory: File)`, which allows us to run a fresh sbt on the given working directory. Next, I've refactored the stateful client-side buffer to a class `TestServer`.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -506,10 +505,9 @@ lazy val sbtProj = (project in file("sbt"))
     buildInfoKeys in Test := Seq[BuildInfoKey](
       // WORKAROUND https://github.com/sbt/sbt-buildinfo/issues/117
       BuildInfoKey.map((fullClasspath in Compile).taskValue) { case (ident, cp) => ident -> cp.files },
+      classDirectory in Compile,
+      classDirectory in Test,
     ),
-    connectInput in run in Test := true,
-    outputStrategy in run in Test := Some(StdoutOutput),
-    fork in Test := true,
```
Comment (dwijnand) [inline build.sbt:512]: losing these breaks `sbtOn`, and the remote debugging was also useful when using `sbtOn`.

**Review by dwijnand — CHANGES_REQUESTED** [review]
these changes impact `sbtOn`'s behaviour

File: build.sbt (hunk 1 above)
Comment (eed3si9n) [reply build.sbt:512]: I made `fork` a function, so I am guessing I can write sbtOn based on that?

Comment (dwijnand) [reply build.sbt:512]: sure.

just make sure: it works, it sends its output immediately to the shell (e.g the prompt text), and it setups the remote debugging port.

Comment (eed3si9n) [reply build.sbt:512]: Fixed sbtOn.

Comment (dwijnand) [reply build.sbt:512]: Thank you!

**Review by dwijnand — APPROVED** [review]

**Pull request merged**